src/modules/ekf2/EKF/output_predictor.cpp

```diff
@@ -48,126 +48,183 @@ void OutputPredictor::print_status()
-		_output_buffer[i].quat_nominal = q_delta * _output_buffer[i].quat_nominal;
-		_output_buffer[i].quat_nominal.normalize();
-		_output_buffer[i].vel += vel_delta;
-		_output_buffer[i].pos += pos_delta;
-	}
-
-	_output_new = _output_buffer.get_newest();
 }
 
 void OutputPredictor::reset()
 {
-	// TODO: who resets the output buffer content?
-	_output_new = {};
-	_output_vert_new = {};
+	_output_buffer.reset();
 
 	_accel_bias.setZero();
 	_gyro_bias.setZero();
 
 	_time_last_update_states_us = 0;
 	_time_last_correct_states_us = 0;
 
+	_output_new = {};
 	_R_to_earth_now.setIdentity();
-	_vel_imu_rel_body_ned.setZero();
-	_vel_deriv.setZero();
-
-	_delta_angle_corr.setZero();
+	_vel_imu_rel_body_ned.zero();
+	_vel_deriv.zero();
 
-	_vel_err_integ.setZero();
-	_pos_err_integ.setZero();
+	_delta_angle_corr.zero();
+	_vel_err_integ.zero();
+	_pos_err_integ.zero();
 
-	_output_tracking_error.setZero();
-
-	for (uint8_t index = 0; index < _output_buffer.get_length(); index++) {
-		_output_buffer[index] = {};
-	}
+	_output_tracking_error.zero();
 
-	for (uint8_t index = 0; index < _output_vert_buffer.get_length(); index++) {
-		_output_vert_buffer[index] = {};
-	}
+	_output_filter_aligned = false;
 }
 
-void OutputPredictor::resetQuaternion(const Quatf &quat_change)
+void OutputPredictor::resetQuaternion(const Quatf &new_quat)
 {
+	const outputSample &output_delayed = _output_buffer.get_oldest();
+	const Quatf quat_change{(new_quat * output_delayed.quat_nominal.inversed()).normalized()};
+
 	// add the reset amount to the output observer buffered data
 	for (uint8_t i = 0; i < _output_buffer.get_length(); i++) {
-		_output_buffer[i].quat_nominal = quat_change * _output_buffer[i].quat_nominal;
+		_output_buffer[i].quat_nominal = (quat_change * _output_buffer[i].quat_nominal).normalized();
+
+		// apply change to velocity
+		_output_buffer[i].vel = quat_change.rotateVector(_output_buffer[i].vel);
+		_output_buffer[i].vel_alternative = quat_change.rotateVector(_output_buffer[i].vel_alternative);
 	}
 
 	// apply the change in attitude quaternion to our newest quaternion estimate
-	// which was already taken out from the output buffer
-	_output_new.quat_nominal = quat_change * _output_new.quat_nominal;
+	_output_new.quat_nominal = (quat_change * _output_new.quat_nominal).normalized();
+	_R_to_earth_now = Dcmf(_output_new.quat_nominal);
+
+	// apply change to velocity
+	_output_new.vel = quat_change.rotateVector(_output_new.vel);
+	_output_new.vel_alternative = quat_change.rotateVector(_output_new.vel_alternative);
+
+	propagateVelocityUpdateToPosition();
 }
 
-void OutputPredictor::resetHorizontalVelocityTo(const Vector2f &delta_horz_vel)
+void OutputPredictor::resetHorizontalVelocityTo(const Vector2f &new_horz_vel)
 {
-	for (uint8_t index = 0; index < _output_buffer.get_length(); index++) {
-		_output_buffer[index].vel.xy() += delta_horz_vel;
+	const outputSample &output_delayed = _output_buffer.get_oldest();
```

Comment:
Is the oldest element of the buffer really the same as the delayed estimate? Isn't it the one predicted one step forward?

Reply:
Reviewing...

Reply:
I said that because of this comment: https://github.com/PX4/PX4-Autopilot/pull/20960/files#r1333296617